lib/Dialect/SPIRV/SPIRVOps.cpp

```diff
@@ -227,6 +227,33 @@ static LogicalResult verifyMemoryAccessAttribute(LoadStoreOpTy loadStoreOp) {
   return success();
 }
 
+template <typename BarrierOp>
+static LogicalResult verifyMemorySemantics(BarrierOp op) {
+  auto memorySemantics = op.memory_semantics();
+  SmallVector<spirv::MemorySemantics, 4> memorySemanticsBits{
+      spirv::MemorySemantics::Acquire, spirv::MemorySemantics::Release,
+      spirv::MemorySemantics::AcquireRelease,
+      spirv::MemorySemantics::SequentiallyConsistent};
+
+  uint32_t bitCount = 0;
+  for (auto &memorySemanticsBit : memorySemanticsBits) {
+    bitCount += static_cast<uint32_t>(
+        spirv::bitEnumContains(memorySemantics, memorySemanticsBit));
+  }
+
+  if (!bitCount) {
+    return op.emitError(
+        "expected at least one constraint for the memory order");
+  }
+
+  if (bitCount > 1) {
```

Comment:
I am not sure of whether this is part of SPIR-V spec itself. It might be for specific use cases of SPIR-V, but if its not part of the spec itself, better to not have this validation.

Reply:
@MaheshRavishankar thanks a lot for review, I was looking to this link: https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_memory_semantics__id_a_memory_semantics_lt_id_gt

> Despite being a mask and allowing multiple bits to be combined, it is invalid for more than one of these four bits to be set: Acquire, Release, AcquireRelease, or SequentiallyConsistent. Requesting both Acquire and Release semantics is done by setting the AcquireRelease bit, not by setting two bits.

> Memory semantics define memory-order constraints, and on what storage classes those constraints apply to. The memory order constrains the allowed orders in which memory operations in this invocation can made visible to another invocation. The storage classes specify to which subsets of memory these constraints are to be applied. Storage classes not selected are not being constrained.

Do I miss something?

Reply:
Yeah, this is from the spec.

Reply:
I think it would be nice to quote and refer to the spec everywhere appropriate as code comment in general. Some part of the clang parser are doing this and this is quite helpful reading code.

Reply:
@joker-eph I'll add the comment with link to the spec.